### DISCUSSION TITLE
麻烦帮忙review，谢谢

### DISCUSSION THREAD

**Discussion by wangsongpower** [BYJRK/DotNet-Discussions#118] [C# 技术交流]
[test.zip](https://github.com/BYJRK/DotNet-Discussions/files/12012431/test.zip)

**Comment by BYJRK** [comment]
精力有限，我一点一点看过去吧。

p2，你不应该去注册 `RadioButton` 的 `Click` 事件，这样你完全就是把它当成 `Button` 来使用了（而且看起来即便你现在已经选中了，再次点击可能依旧有效？）；
你应该考虑用一个 `Converter` 将两个 `RadioButton` 的 `IsChecked` 属性绑定到一个属性之类，然后将被控制的 `ViewModel` 作为一个可以绑定的属性，并被 `RadioButton` 切换。

p3，你将引用设为 `null` 并不会让它被释放，因为首先即便 VM 真的已经没有被引用了，GC 通常也并不会立刻去释放（除非你调用 `GC.Collect()`）；其次，VM 完全有可能因为各种方式被引用到，比如事件注册等。所以你可以考虑让两个 VM 不被释放，而是持久化地存在 `MainViewModel` 中，然后每次也不必 new 一个出来。

p4，弱引用并不需要你去主动释放，因为它并不会影响对象被释放，而且你写在析构函数里的方法的被调用时机也很不可控。另外，如果你有类似的需求，应该考虑继承 `ObservableRecipient` 类，它原生提供了很多相关的功能。

**Reply by wangsongpower** [reply]
非常感谢抽空回复。
p2，是的，完全当作button去用的，再次点击的话，我会去做一个判断的，另外converter还不怎么会用。
p3，我还是需要直接释放vm，而不是都放进内存里，网上找了好多，现在能想到的只有disposable，不知道如何具体实现。
p4，好的，正好刚看了您的视频。

如果可以的话，还是希望有个简单的sample code，或者直接丢给我一个link或者小项目，我自己去研究，谢谢

**Reply by BYJRK** [reply]
p2 你可以参考我的视频：https://www.bilibili.com/video/BV1E14y137qd
p3 你可以去了解一下 IDisposable 接口的实现及使用

**Comment by BYJRK** [comment]
p5，你每次切换都会实例化一个新的 `BitmapImage`，这是不合理的。应该考虑存成静态字段
p6，如果对应的控件是一个 `Button`，那么直接绑定 `Command` 属性即可；否则恐怕只能使用 `EventTrigger` 了
p7-8，我没太看懂你的需求
p9，你的业务逻辑有点复杂，我没怎么理清楚。不过写 UC 一般需要注意的就是如何将 `View` 与 `ViewModel` 进行绑定；或者你也可以不绑定，而是写依赖属性，再在前台实例化 UC 的地方对相关属性进行绑定
p10，你的 `Image` 应该避免后台的实例化，因为这可能需要你去手动 `Dispose()`；简易的方法是写成静态变量或资源进行引用，也可以写成 `Style`，让前台根据 `Trigger` 去自动加载，这样通常并不需要你去操心图片资源的释放问题

**Reply by wangsongpower** [reply]
再次感谢，最近我又去研究了不少，学习了英文播主的一个方法。
p2，对于radiobutton，之后我打算bind他的check和uncheck事件，这样就不用判断重复点击了吧？
至于导航，现在采用dependency injection，v与vm的关联在app.xaml利用datatemplate，IServiceCollection与ServiceProvider进行导航，然后virtual一个dispose用来释放vm，v则是自动gc拾取。
p5，我也发现了这个问题，现在用的是要么xaml内直接引用资源，要么vm里bgeininit endinit，赋值给imagesource后，freeze，dispose时给source freeze然后null。
p7-8，这里主要是想实现一个二级，三级导航，就是某个页面内，最左边有个L1导航，然后进去后，里面【可能】还有L2与L3，当点击某个L1会出现L2，点击某个L2会出现L3，然后还有一个上一页下一页功能，当某个页面不可用（按钮为IsEnabled=false）时自动跳过，导航到之后的页面。
我的做法是把他们的IsCheckked，IsEnabled的都绑定，然后制作一个集合，把可用页面的index都放进去，前一页后一页的时候，直接读可用页面里的index。
p9，确实有点恶心，抱歉了，简单来说，有一个控件UC，里面有datagrid，然后里面又有button，点击button后弹出一个窗口，再把窗口里的返回值传给调用uc的vm，值的流向为win→datagrid→uc→vm，非常绕，现在想想，在win里面弄一个weakmessenger会更好吧？
p10，明白了，有的时候业务需求，从硬盘读image，自己dispose吧。如果把Model class（里面有image或者List<>），add进List或者ObservableCollection，需要先把image或者List释放后再把List或者ObservableCollection Clear吧？
